### PR TITLE
testutils refactors and cleanup

### PR DESCRIPTION
**Pull request by smarshall-spitzbart**
Corrected version of https://github.com/cosmos/interchain-security/pull/467

This PR should not change test behavior in anyway. It should allow the implementation of multiple consumers in e2e tests to be cleaner. Changes included: 
- Misc cleanup of testing documentation
- Remove the unused `testutil/sample` folder
- Clean up what was `testutil/simapp/simapp.go` and is now `testutil/ibc_testing/specific_setup.go` `testutil/ibc_testing/generic_setup.go` with semantics closer to the design of the IBC testing package. I've also made those files more succint, and less boilerplate-y
- Adjust e2e test instantiation to use the new IBC testing util functions

### CODE REVIEW

File: tests/e2e/democracy.go [hunk 1]
```diff
@@ -7,15 +7,14 @@ import (
 	"time"
 
 	sdk "github.com/cosmos/cosmos-sdk/types"
-
 	ibctesting "github.com/cosmos/ibc-go/v3/testing"
+	ibctestingutils "github.com/cosmos/interchain-security/testutil/ibc_testing"
```
Comment (MSalopek) [inline tests/e2e/democracy.go:11]: Maybe alias as `icstestingutils`, `icsibctestingutils`. `icsutilsibc`

Can't think of anything better but think that differentiating from `ibctesting` is useful

File: tests/e2e/instance_test.go [hunk 2]
```diff
@@ -3,57 +3,43 @@ package e2e_test
...
-			e2etestutil.ConsumerApp,
-		) {
-			// Here we pass the concrete types that must implement the necessary interfaces
-			// to be ran with e2e tests.
-			coord, prov, cons := simapp.NewProviderConsumerCoordinator(t)
-			return coord, prov, cons, prov.App.(*appProvider.App), cons.App.(*appConsumer.App)
-		},
-	)
+	// Pass in concrete app types that implement the interfaces defined in /testutil/e2e/interfaces.go
+	ccvSuite := e2e.NewCCVTestSuite[*appProvider.App, *appConsumer.App](
```
Comment (MSalopek) [inline tests/e2e/instance_test.go:24]: Nice cleanup!

File: tests/e2e/setup.go [hunk 3]
```diff
@@ -29,28 +29,54 @@ type CCVTestSuite struct {
...
+		coordinator := ibctesting.NewCoordinator(t, 0)
+
+		// Add provider to coordinator, store returned test chain and app.
+		// Concrete provider app type is passed to the generic function here.
+		provider, providerApp := ibctestingutils.AddProvider[Tp](
+			coordinator, t, providerAppIniter)
+
+		// Add specified number of consumers to coordinator, store returned test chains and apps.
+		// Concrete consumer app type is passed to the generic function here.
+		consumers, consumerApps := ibctestingutils.AddConsumers[Tc](
```
Comment (MSalopek) [inline tests/e2e/setup.go:62]: Nice!

File: testutil/ibc_testing/generic_setup.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+	coordinator.Chains[provChainID] = provider
+
+	return provider, provider.App.(T)
+}
+
+// AddDemocracyConsumer adds a new democ consumer chain to the coordinator and returns the test chain and app type
+func AddDemocracyConsumer[T e2e.DemocConsumerApp](coordinator *ibctesting.Coordinator, t *testing.T,
+	appIniter ibctesting.AppIniter) (*ibctesting.TestChain, T) {
+
+	democConsumer := ibctesting.NewTestChain(t, coordinator, appIniter, democConsumerChainID)
```
Comment (MSalopek) [inline testutil/ibc_testing/generic_setup.go:33]: Why is democConsumer setup different from initialization done in `AddConsumers`?

Provider's valset is copied to each of the instantiated consumers so I'm a bit confused. Isn't democracy chain intended to be used as a CCV consumer?

**Review by MSalopek — APPROVED** [review]
LGTM!

A bit confused about the democracy setup, but don't see any issues with the PR

File: tests/e2e/democracy.go (hunk 1 above)
Comment (smarshall-spitzbart) [reply tests/e2e/democracy.go]: Agreed, there's not a super intuitive name that I can think of. But I've at least changed the alias to `icstestingutils`

File: testutil/ibc_testing/generic_setup.go (hunk 4 above)
Comment (smarshall-spitzbart) [reply testutil/ibc_testing/generic_setup.go:33]: It's notable that the democracy consumer e2e tests are [separated from](https://github.com/cosmos/interchain-security/blob/4c96805c51192a4ce9f104119d30400e4d122fe2/tests/e2e/democracy.go) the normal e2e tests. The democracy consumer tests do not run with a provider, so they need to be setup slightly differently (no copying of provider val set).

The democracy consumer should be able to pass the normal set of e2e ccv tests tho, it's not currently able to do this. see https://github.com/cosmos/interchain-security/issues/397

**Review by danwt — APPROVED** [review]

**Pull request merged**